### PR TITLE
Add FontFallback and TextAnalysisSource objects

### PR DESCRIPTION
**Pull request by raphlinus**
Adds wrapping for FontFallback and TextAnalysisSource, including custom implementations of the latter in Rust.

### CODE REVIEW

**Comment by raphlinus** [comment]
I haven't tested this yet, but wanted to get feedback before I spend too much more time on it.

I've found a number of issues in dwrote. For one, the use of unsafe seems quite wrong to me. For example `Font::take` doesn't validate its pointer arg, and among other things it's quite easy for "safe" code to pass a null pointer. Conversely, `as_ptr` is listed as unsafe although it just returns a pointer.

There are also some opportunities to reduce allocations by using smallvecs and the like.

I've tried to make this wrapper pretty thin, but have made some choices. For one, I'm just using a Vec<u16> for the text. Trying to wire up access to non-contiguous buffers would be painful and I'm not convinced would be a performance win. For two, I'm only allowing a single number substitution (this follows Chromium).

I'll update the PR when I have it working end-to-end, but will happily accept feedback meanwhile.

**Comment by raphlinus** [comment]
I have some [example code](https://gist.github.com/raphlinus/7b669e6af16e69877da3a5287c47a506) that seems to work. I can add it under the examples directory, which seems not to exist.

There are possibly some other issues. @connicpu (who is adding similar functionality to directwrite-rs) suggests that the text analysis methods return &[u16] in general. This would allow utf-16 ropes for the text, but, perhaps more to the point, would have less concerns about safety because none of the buffers would be mutable. I'm certainly open to that change.

I'll continue prototyping with this but am open to more suggestions.

File: src/font_fallback.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+impl FontFallback {
+    pub fn get_system_fallback() -> Option<FontFallback> {
+        unsafe {
+            let factory = ComPtr::already_addrefed(DWriteFactory());
+            let factory2 = factory.query_interface::<IDWriteFactory2>(&IDWriteFactory2::uuidof());
+            std::mem::forget(factory);
+            let factory2 = factory2?;
+            let mut native = null_mut();
+            let hr = factory2.GetSystemFontFallback(&mut native);
+            assert!(hr == 0);
```
Comment (pcwalton) [inline src/font_fallback.rs]: nit: could be `assert_eq!()`, that way it'll dump the error code if it fails

File: src/font_fallback.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                base_font.as_ptr(),
+                base_family.map(|s| s.to_wide_null().as_mut_ptr()).unwrap_or(null_mut()),
+                base_weight.t(),
+                base_style.t(),
+                base_stretch.t(),
+                &mut mapped_length,
+                &mut font,
+                &mut scale,
+            );
+            assert!(hr == 0);
```
Comment (pcwalton) [inline src/font_fallback.rs]: ditto

File: src/text_analysis_source_impl.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+use winapi::um::unknwnbase::{IUnknown, IUnknownVtbl};
+use winapi::um::winnt::HRESULT;
+
+use helpers::ToWide;
+use super::DWriteFactory;
+use com_helpers::{Com, UuidOfIUnknown};
+use comptr::ComPtr;
+
+/// The Rust side of a custom text analysis source implementation.
+pub trait TextAnalysisSourceImpl {
```
Comment (pcwalton) [inline src/text_analysis_source_impl.rs]: nit: Kind of odd to name a trait `Impl`. Sometimes I use the word `Methods` for this, like `TextAnalysisSourceMethods`…

File: src/text_analysis_source_impl.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    {
+        unsafe {
+            let mut native: ComPtr<IDWriteNumberSubstitution> = ComPtr::new();
+            let hr = (*DWriteFactory()).CreateNumberSubstitution(
+                subst_method,
+                locale.to_wide_null().as_ptr(),
+                if ignore_user_overrides { TRUE } else { FALSE },
+                native.getter_addrefs(),
+            );
+            assert!(hr == 0, "error creating number substitution");
```
Comment (pcwalton) [inline src/text_analysis_source_impl.rs]: nit: same, use `assert_eq!()`

File: src/text_analysis_source_impl.rs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    GetTextBeforePosition: CustomTextAnalysisSourceImpl_GetTextBeforePosition,
+};
+
+impl CustomTextAnalysisSourceImpl {
+    /// Create a new custom TextAnalysisSource for the given text and a trait
+    /// implementation.
+    ///
+    /// Note: this method only supports a single `NumberSubstitution` for the
+    /// entire string.
+    pub fn new_native(inner: Box<dyn TextAnalysisSourceImpl>, text: Vec<wchar_t>,
```
Comment (pcwalton) [inline src/text_analysis_source_impl.rs]: I don't love only supporting a part of the interface here, but I understand doing this for expedience' sake. My only inclination in this case would be to name the methods in such a way as to allow exposing the API in its full generality in the future. For example, we could have something like `from_text_and_number_subst`, reserving the name `new` for the full API in the future. What do you think?

File: src/font_fallback.rs (hunk 1 above)
Comment (raphlinus) [reply src/font_fallback.rs]: I'm following crate conventions here (the pattern is used extensively). I'm not a fan in any case - it's panic-prone when things do go wrong. Do you want me to use assert_eq in my new code only (which would make it inconsistent with rest of crate), or change all instances? Or leave it alone?

File: src/text_analysis_source_impl.rs (hunk 3 above)
Comment (raphlinus) [reply src/text_analysis_source_impl.rs]: Sure, I was pretty unsure about this.

File: src/text_analysis_source_impl.rs (hunk 5 above)
Comment (raphlinus) [reply src/text_analysis_source_impl.rs]: Sure.

File: src/font_fallback.rs (hunk 1 above)
Comment (pcwalton) [reply src/font_fallback.rs]: Ah, I see. I guess using `assert_eq!` in new code only is fine.

**Pull request merged**